Refactor list parser

wouterj opened the pull request:
Fixes #142, fixes https://github.com/weaverryan/docs-builder/issues/70, fixes https://github.com/weaverryan/docs-builder/issues/38

Previously, lists were parsed as flat structure and the renderer took care of finding out the nested lists. That didn't really work and didn't allow subparsing the list item (e.g. directives or code blocks nested in the list). This PR refactors it to parse lists using a subparser, similarly to definition lists and directives.

I also added a little `tests/sphinx` utility that allows you to easily run the functional using a locally installed Sphinx: `./tests/sphinx list` will render `tests/Functional/tests/list/list.rst`. All tests in this PR are validated using Sphinx (there were quite some tests that didn't really follow the official specs).

Todo
---

I wanted to get this PR ready for review as soon as possible. A few things needs to be done:

* [x] There are BC breaks in here when using the PHP classes, probably document them in `UPGRADE.md`
* [x] I don't know LaTex really well, it's output currently is extremely broken (also due to Twig whitespace control)
* [x] Test this branch on weaverryan/docs-builder, to get an extra set of eyes on whether this breaks important things.  
    *Done, no changes needed and all green*

wouterj committed:
Generalize some LineChecker methods

wouterj committed:
Completely rewrite list parsing

This refactors parsing lists using subparsers, in order to parse list contents
(like code blocks or other body elements).

It also fixed some inconsistencies between the reStructuredText Markup
specification and the behavior of this parser.

wouterj committed:
Replace escape characters in generated output

wouterj committed:
Add utility to easily run Sphinx on fucntional tests

wouterj commented on tests/Functional/tests/enumerated-ignored/enumerated-ignored.rst:
```diff
@@ -0,0 +1,6 @@
+The second line of each enumerated list item is checked for validity:
+
+1. This is not a list.
+It's a paragraph starting with a number.
```
Previously, there was a tests that tested this behavior as "being recognized as a list" (`tests/Functional/tests/ordered2/ordered2.rst`). However, [the markup specs](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#enumerated-lists) are very explicit about this:

> The second line of each enumerated list item is checked for validity. This is to prevent ordinary paragraphs from being mistakenly interpreted as list items, when they happen to begin with text identical to enumerators. For example, this text is parsed as an ordinary paragraph:
> 
> ```
> A. Einstein was a really
> smart dude.
> ```
> 
> However, ambiguity cannot be avoided if the paragraph consists of only one line. This text is parsed as an enumerated list item:
> 
> `A. Einstein was a really smart dude.`
> 
> If a single-line paragraph begins with text identical to an enumerator ("A.", "1.", "(b)", "I)", etc.), the first character will have to be escaped in order to have the line parsed as an ordinary paragraph:
> 
> `\A. Einstein was a really smart dude.`

wouterj commented on tests/Functional/tests/escape/escape.rst:
```diff
@@ -1 +1,3 @@
 Testing that this is escaped: <script>
+
+\And escape \characters are ig\nored, but this one isn't: \\stdClass
```
Escape characters are not rendered according to the Sphinx parser. This was also required to make `\1. Single line paragraphs starting with a number may use an escape.` pass.

wouterj commented on tests/Functional/tests/indented-list/indented-list.html:
```diff
@@ -1,7 +0,0 @@
-<p>Testing an indented list:</p>
```
This tested indented lists to be parsed as lists. But in fact, reSt doesn't allow any whitespace before the list marker. Such whitespace (indentation) is recognized as a blockquote.

This is now tested by `tests/Functional/tests/list-indented/list-indented.rst`

wouterj commented on tests/Functional/tests/standalone-email-addresses/standalone-email-addresses.html:
```diff
@@ -79,8 +79,15 @@ <h1>
     <hr />
-    <p>Received: from x.y.test by example.net via TCP with ESMTP id ABC12345 for &lt;<a href="mailto:mary@example4.net">mary@example4.net</a>&gt;; 21 Nov 1997 10:05:43 -0600
-Received: from node.example by x.y.test; 21 Nov 1997 10:01:22 -0600
+    <dl>
```
Honestly, I'm blown away by this test file. It was added to test the automatic recognition of e-mailaddresses. However, testing raw emails headers as reStructuredText seems very weird (resulting in strange things like this definition list).

wouterj commented on lib/Parser/LineChecker.php:
```diff
@@ -43,15 +37,18 @@ public function isSpecialLine(string $line): ?string
         return $letter;
     }
 
-    public function isListLine(string $line, bool $isCode): bool
```
I'm not sure what the `$isCode`'s function was here. Code blocks shouldn't be subparsed, so this should never be true. So I simply removed it.

weaverryan replied:
The only place it was used is in DocumentParser... related to lists... and you have already refactored that part. So this is fine with me - probably some old, unwanted piece related to the odd "is code" handling in DocumentParser.

wouterj commented on tests/Functional/tests/list-alternate-syntax/list-alternate-syntax.rst:
```diff
@@ -2,10 +2,17 @@ Alternate List Syntax
+-
+    But with a more indented first line,
+    this same level is a normal paragraph
+  and this is not part of the list (warning)
```
Sphinx triggers a warning here. Can we trigger warnings in the parser for cases like this? (lists have to end with a blank line)

greg0ire replied:
If Sphinx does, I think we can.

wouterj commented on lib/Parser/LineChecker.php:
```diff
@@ -6,21 +6,15 @@
-    {
-        $this->lineParser = $lineParser;
-    }
+    public const LIST_MARKER     = '/^([-+*\x{2022}\x{2023}\x{2043}]|(?:[\d#]+\.|[\d#]+\)|\([\d#]+\)))(?: +|$)/u';
```
This is btw not a complete set of allowed markers. Enumerated lists can also use `A-Z`, `a-z` and even Roman `I-MMMMCMXCIX`. In order to not make this library overly complex, I only implemented the numeric markers (which was also the only one supported previously by this library).

greg0ire replied:
Let's use the `x` modifier and add comments to that one? :sweat_smile:

greg0ire commented on lib/Span/SpanProcessor.php:
```diff
@@ -288,6 +290,11 @@ private function replaceStandaloneEmailAddresses(string $span): string
         );
     }
 
+    public function replaceEscapes(string $span): string
```
Why is that method public?

greg0ire commented on lib/Span/SpanProcessor.php:
```diff
@@ -288,6 +290,11 @@ private function replaceStandaloneEmailAddresses(string $span): string
 
+    public function replaceEscapes(string $span): string
+    {
+        return preg_replace('/(?<!\\\\)\\\\/', '', $span);
```
I don't understand that regex, but that's probably because I don't know what the list of characters you want to replace is, and what are actually metacharacters in that regex. Can you please tell us what is the list?

wouterj replied:
We want to replace `\` in normal text, but we need to keep a backslash when it's escaped (`\\`).

So I match every backslash that isn't preceded by another backslash: `(?<!\)\`. And the joy of PHP strings escapes and regex escapes means that we need 4 backslashes to match a single backslash (`'\\\\'` is transformed into `\\` by PHP, and into `\` by the regex engine).

greg0ire replied:
> every backslash that isn't preceded by another backslash

Oh ok 👍 let's add this in a code comment?

greg0ire commented on lib/HTML/Renderers/ListNodeRenderer.php:
```diff
@@ -0,0 +1,34 @@
+use Doctrine\RST\Renderers\NodeRenderer;
+use Doctrine\RST\Templates\TemplateRenderer;
+
+class ListNodeRenderer implements NodeRenderer
```
Let's make it `final` as well as every other new class?

greg0ire commented on lib/Nodes/ListNode.php:
```diff
@@ -6,29 +6,33 @@
+    protected $ordered;
+
+    /** @var Node[] */
+    protected $items;
```
Let's make things `private` by default?

wouterj replied:
I'm not sure, all nodes are using protected properties. I think it's best to change it all in 1 go if you would like that (and honestly, this PR is already much bigger than I wanted, so I prefer to not do it in this PR)

greg0ire replied:
Here we have both: https://github.com/doctrine/rst-parser/blob/9f9887b282307c7cc8ff92a2c98aec39346696d9/lib/Nodes/Node.php#L25-L37

If we are going to migrate to private in the future, let's make it less work? That way we will not have to check if there are references to these 2 properties in extending classes.

greg0ire replied:
Addressed in https://github.com/doctrine/rst-parser/pull/144

wouterj commented:
Updated this PR. Fixed behavior when mixing different list markers and fixed the LaTex out (and checked its validness with a LaTex parser).

<s>Later today, I'm going to reintroduce `list-item.html.twig` (and it's usage in both list templates). That would significantly reduce the impact of this PR on its users (this change currently breaks weaverryan/docs-builder).</s> done

weaverryan commented on lib/Nodes/ListNode.php:
```diff
@@ -4,31 +4,37 @@
+     * @param ListItem[] $list
      */
-    public function addLine(array $line): void
+    public function __construct(array $list, bool $ordered)
```
```suggestion
    public function __construct(array $items, bool $ordered)
```

weaverryan reviewed (approved):
This PR is terrifying... in a good way! 

This does break some BC... but all in ways to, clearly, drastically fix the list rendering. A bug 👍 from me!

wouterj commented:
This PR should be ready. I've documented the BC breaks very minimalistic, I don't really expect anyone to be hit by these.

greg0ire commented:
I tried the new utility locally, it errors out like this:

<details>

```

Running Sphinx v3.2.1
making output directory... done
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 75 source files that are out of date
updating environment: [new config] 75 added, 0 changed, 0 removed
reading sources... [100%] wrap/wrap
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/anchor-failure/anchor-failure.rst:1: WARNING: Unknown target name: "@anchor section".
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/comment-3/comment-3.rst:2: WARNING: Explicit markup ends without a blank line; unexpected unindent.
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/css/css.rst:1: WARNING: Unknown directive type "stylesheet".

.. stylesheet:: style.css
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/div/div.rst:1: WARNING: Unknown directive type "div".

.. div:: testing
    Hello!
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/figure/figure.rst:4: WARNING: image file not readable: figure/foo.jpg
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/image-inline/image-inline.rst:3: WARNING: image file not readable: image-inline/test.jpg
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/image/image.rst:5: WARNING: Error in "image" directive:
unknown option: "title".

.. image:: other.jpg
    :title: Other
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/image/image.rst:1: WARNING: image file not readable: image/test.jpg
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/image/image.rst:: WARNING: image file not readable: image/try.jpg
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/list-alternate-syntax/list-alternate-syntax.rst:18: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/list-mix/list-mix.rst:2: WARNING: Bullet list ends without a blank line; unexpected unindent.
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/list-mix/list-mix.rst:3: WARNING: Bullet list ends without a blank line; unexpected unindent.
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/main-directive/main-directive.rst:3: WARNING: Unknown directive type "latex-main".

.. latex-main::
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/pretty-table-error1/pretty-table-error1.rst:8: WARNING: Blank line required after table.
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/quote3/quote3.rst:2: WARNING: image file not readable: test.jpg
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/raw/raw.rst:3: WARNING: Error in "raw" directive:
1 argument(s) required, 0 supplied.

.. raw::

    <u>Underlined!</u>
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/simple-table-error1/simple-table-error1.rst:3: WARNING: Malformed table.
Text in column margin in table line 3.

=========== ========== =========
First col   Second col Third col
Second row  Other colllOther col
Third  row  Other col  Last col
=========== ========== =========
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/standalone-email-addresses/standalone-email-addresses.rst:5: WARNING: Unexpected section title or transition.

----
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/standalone-email-addresses/standalone-email-addresses.rst:19: WARNING: Unexpected section title or transition.

----
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/standalone-email-addresses/standalone-email-addresses.rst:27: WARNING: Title underline too short.

Hi everyone.
----
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/standalone-email-addresses/standalone-email-addresses.rst:27: WARNING: Unexpected section title.

Hi everyone.
----
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/standalone-email-addresses/standalone-email-addresses.rst:31: WARNING: Unexpected section title or transition.

----
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/standalone-email-addresses/standalone-email-addresses.rst:39: WARNING: Title underline too short.

Testing.
----
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/standalone-email-addresses/standalone-email-addresses.rst:39: WARNING: Unexpected section title.

Testing.
----
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/standalone-email-addresses/standalone-email-addresses.rst:43: WARNING: Unexpected section title or transition.

----
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/standalone-email-addresses/standalone-email-addresses.rst:53: WARNING: Title underline too short.

This is a reply to your reply.
----
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/standalone-email-addresses/standalone-email-addresses.rst:53: WARNING: Unexpected section title.

This is a reply to your reply.
----
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/standalone-email-addresses/standalone-email-addresses.rst:57: WARNING: Unexpected section title or transition.

----
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/standalone-email-addresses/standalone-email-addresses.rst:74: WARNING: Unexpected section title or transition.

----
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/standalone-email-addresses/standalone-email-addresses.rst:81: WARNING: Definition list ends without a blank line; unexpected unindent.
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/standalone-email-addresses/standalone-email-addresses.rst:97: WARNING: Unexpected section title or transition.

----
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/standalone-email-addresses/standalone-email-addresses.rst:100: WARNING: Unexpected indentation.
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/standalone-email-addresses/standalone-email-addresses.rst:102: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/standalone-email-addresses/standalone-email-addresses.rst:103: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/standalone-email-addresses/standalone-email-addresses.rst:105: WARNING: Unexpected indentation.
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/standalone-email-addresses/standalone-email-addresses.rst:110: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/standalone-email-addresses/standalone-email-addresses.rst:113: WARNING: Title underline too short.

Testing.
----
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/standalone-email-addresses/standalone-email-addresses.rst:113: WARNING: Unexpected section title.

Testing.
----
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/standalone-hyperlinks/standalone-hyperlinks.rst:91: WARNING: Unknown target name: "blah".
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/unknown-directive/unknown-directive.rst:1: WARNING: Unknown directive type "unknown-directive".

.. unknown-directive::
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/variable-wrap/variable-wrap.rst:1: WARNING: Substitution definition "test" empty or invalid.

.. |test| note::
    This is an important thing
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/variable-wrap/variable-wrap.rst:4: WARNING: Undefined substitution referenced: "test".
/home/greg/dev/doctrine-rst-parser/tests/Functional/tests/variable-wrap/variable-wrap.rst:6: WARNING: Undefined substitution referenced: "test".

Sphinx error:
master file /home/greg/dev/doctrine-rst-parser/tests/Functional/tests/.rst not found

```

</details>

I was looking into it because I wanted to understand if the new gitignored directory was going to be useful or if we should maybe use `/tmp` instead, thus avoiding `rm -rf` and a new gitignore entry

wouterj commented:
@greg0ire I intended the new utility to be used like `./tests/sphinx list-alternate-syntax` (so for the specific functional test you want).

> I was looking into it because I wanted to understand if the new gitignored directory was going to be useful or if we should maybe use /tmp instead, thus avoiding rm -rf and a new gitignore entry

I put it in the project dir because I want to look at the generated HTML (that's the whole purpose of running Sphinx). Opening a file from the project dir is much easier (at least for me) than opening a tmp directory.

greg0ire commented:
Thanks @wouterj !

wouterj commented:
@greg0ire thanks for merging :)

For your information, I've just opened #145 - which is the open comment from Ryan's PR. I don't think it's feasible to fix that in any very short future, so if you would like to draft a new release with these 2 big PRs - feel free to do so :)
